File: tmb_examples/TMBad/spa_gauss.R
## Simulate data
set.seed(123)
n <- 1000
mu <- 3
sigma <- 1.5
y <- rnorm(n, mu, sigma)

## data and parameters
data <- list(y=y,s=numeric(n))
parameters <- list(mu=0, logSigma = log(1))

## Compile C++ code and load into R
library(TMB)
compile("spa_gauss.cpp", framework="TMBad")
dyn.load(dynlib("spa_gauss"))

## create adfun (SPA inner problem option set in C++ template)
obj <- MakeADFun(data, parameters, DLL="spa_gauss")

## optimize
opt <- nlminb(obj$par, obj$fn, obj$gr)
rep <- sdreport(obj)
summary(rep, p.value=TRUE)

## Check values of nll vs nlspa
obj$fn(opt$par) 
-sum(dnorm(y, opt$par[1], exp(opt$par[2]), log=TRUE))

## Check values of dnll vs dnlspa
obj$gr(opt$par)
c(
    sum((opt$par[1]-y)/exp(opt$par[2])^2), # dnll / dmu
    -sum(exp(-2*opt$par[2])*(y-opt$par[1])^2) + n # dnll / dlogsigma
)

## check s = normal sp and parameters
##   arg min K - sx
##   K' - x = 0
##   K' = mu + sigma^2 s = x
##   s = (x - mu)/sigma^2
s_hat <- obj$report()$s
plot(s_hat, (y-opt$par[1])/(exp(opt$par[2]))^2,
     main="Numerical versus theoretical saddlepoints",
     xlab="Numerical", ylab="Theoretical")

File: tmb_examples/TMBad/spa_gauss.cpp
// Demonstrate saddlepoint approximation (SPA)
#include <TMB.hpp>

// Define the ad type needed
typedef TMBad::ad_aug ad;

struct spa_gauss {
  vector<ad> y; // Data
  ad mu, sigma; // Parameters
  // Return K_y(s)-s^T y
  // K_y(s) = sum K_{y[i]}(s[i])
  // K_N(mu, sigma)(s) = mu*s + sigma^2 * s^2 / 2
  ad operator()(vector<ad> s) {
    int n = s.size();
    // Build CGF
    ad K = 0;
    for(int i=0; i<n; i++){
      K += mu * s(i) + 0.5 * s(i) * s(i) * sigma * sigma;
    }
    // Build inner problem
    ad res = K - (s*y).sum();
    return res;
  }
  // Type can by 'ad' or 'double'
  template<class Type>
  Type eval_nldens(vector<Type> &start) {
    newton::newton_config cfg;
    cfg.SPA = true;
    cfg.sparse = true;
    cfg.trace  = false; // Trace inner optimize?
    Type res = newton::Laplace(*this, start, cfg);
    return res;
  }
};

template<class Type>
Type objective_function<Type>::operator() ()
{
  DATA_VECTOR(y);
  DATA_VECTOR(s); // saddlepoint initial guess
  PARAMETER(mu);
  PARAMETER(logSigma);
  Type sigma = exp(logSigma);

  spa_gauss spa = {y, mu, sigma};

  // s = start guess on input
  // overwritten with solution on output
  Type res = spa.eval_nldens(s);

  // report saddlepoint and sigma
  REPORT(s); // sd not needed
  ADREPORT(sigma);

  return res;
}
